README.markdown
Ultima IV Remastered
====================

This is the source for my remastered C64 port of Ultima IV. Details on what has been improved over the original can be found [on my blog](http://magervalp.github.io/2015/03/30/u4-remastered.html).


Required Tools
--------------

To build the source you need:

* [cc65](https://github.com/cc65/cc65)
* c1541 from [VICE](http://sourceforge.net/projects/vice-emu/)
* [exomizer](http://hem.bredband.net/magli143/exo/)
* Python 2.x
* GNU Make


Building
--------

To build the source you need disk images of the original Ultima IV release placed in `files/original`. Untouched original images should have the following sha1 checksums:

    86e118fb496e9e7b5126bc0d1c8fa62129f232fd  files/original/u4britannia.d64
    334228897ff774537e56ad706c0ede3afcf7ebc1  files/original/u4program.d64
    728b6cb40ea5dac9e24c31084d70c852ad7eae24  files/original/u4towne.d64
    312984d7aa7a0b8df40003a5852e8827c436c724  files/original/u4underworld.d64

Track 6 on the Program disk contains errors as part of the copy protection and the checksum can differ depending on which tool is used to create `u4program.d64`. The checksum for `u4britannia.d64` also differs if the Britannia disk has been saved to, but as long as both are made from original disks the source will build without problems.

The source then builds from the top Makefile, e.g. with:

    make -j 8


License and Copyright
---------------------

The code is &copy; 2006-2015 Per Olofsson, and is available under an Apache 2.0 license.

The new graphics are &copy; 2015 Vanja Utne.

The crack intro music is &copy; 2015 Johan Samuelson.

The original Ultima IV game is &copy; 1985 Origin Systems, Inc. However, no data from the original game is included in this repository.


Changelog.txt
# Ultima IV Remastered Changelog

## v2.4.1 (2021-10-22)

### Critical Bugs

* Fix graphics corruption in endgame leading to crash. Introduced with 2.4.0

### Other Errors

* Two ladders in dungeon "Wrong" are mis-matched.
* Dungeon "Covetous" has a room whose east exit is obstructed.
* Cooldown timer for Hawkwind's virtue reward should match timer for shrine.
* Cooldown timer for reagent searches should not reset when search finds non-reagents.


## v2.4.0 (2021-09-05)

More fixes by Sean Gugler.

### Critical Bugs

* Stack overflow when ship sinks with player aboard.

### Other Errors

* Clear keyboard in Shrine mis-copied from Apple II.
* Better NPC choice for restoring Thevel - pick the unmoving Joe, not one of the mobile ones.
* Skull use in battlefield should affect combatants, not clear outer world of wandering monsters.
* Skull use in dungeon only immobilizes wandering monsters, doesn't kill them.
* Skull cannot be used in *any* dungeon rooms; fixed to behave normally except in abyss, where it is disallowed entirely.
* Player can be attacked by bridge trolls while drifting across in the balloon.
* Dismounting horse in certain townes causes a random NPC to vanish.
* Chests dropped by townespeople can be talked to.

### Cosmetic Mistakes

* Using skull in Britannia is silent.
* More text revisions.
* Ankh doesn't immediately update after attacking townsfolk and losing virtues.
* Animation flicker when attacking an east-facing horse.
* Original graphics for energy fields restored.

### Enhancements

* Gender question did not support non-binary; add a neutral symbol.
* Library shelf lettering made one puzzle somewhat awkward.
* More source code: end_game.s
* Hard-coded string table offset in Shrine replaced with symbol.
* Allow player to initiate attack in dungeon, after using trainer to avoid.
* Teleport trainer only wipes monsters from map, preserve any chests and transports.
* Proper descriptions for non-human NPCs instead of "phantom".
* Campfire NPC remains stationary in combat and spits fire.
* Ctrl+V turns sound effects on and off during play.
* Balloon trainer can klimb twice to drift, descend to steer again.


## v2.3.0 (2021-03-25)

* Added fixes from Sean Gugler for the following issues.

### Critical Bugs

* Stack overflow every moongate jump.
* Stack overflow using Ztats during combat.
* Stack overflow using Z-down spell at bottom of dungeon.
* Stack underflow if characters attempt to take different exits from dungeon room.
* Morning after inn wakes sleeping characters with invalid status 'A'.

### Other Errors

* Cannot talk to the healer in Serpent's Hold castle.
* Enemy strength for non-monsters was calculated wrong.
* Completion of skull quest grants awards to wrong area of memory.
* NPC questions worded as humility tests were not flagged to affect virtue: Browning, Aristotle, Santayana, Slysam.
* Following advice from cycle 1 at shrine of sacrifice mistakenly awards honor points.
* Trainer improperly initializes "Idle" cheat.

### Cosmetic Mistakes

* Katrina prints both excuses if virtue is too low ("Thou art not humbleexperienced enough").
* Ztats display of 3-part key is truncated if you have bell or book but not candle.
* Spells cast by healer shopkeepers are supposed to have the display inversion effect.
* Spelling, capitalization, and grammar.
* Added more disassembly to fix text issues that can't be patched in place.

### Enhancements

* Additional torch brightness level in dungeon, full white as in original game.
* Trainer for fair buy-back prices.
* Documented zero-page flag "last_meditated".


## v2.2.1 (2015-05-29)

* Fixed graphics corruption and potential crash during ending, thanks Trurl!


## v2.2 (2015-04-22)

* The game no longer hangs when playing the 1541 version and answering questions wrong in the codex.
* Attacking static NPCs such as Ankhs, castle lords & ladies, the water in Castle Britannia, or the camp fire in Yew no longer leads to a glitching enemy in combat. Attacking chests dropped by townespeople is no longer possible.
* Fixed stack handling for main menu, when being attacked by monsters, when dying, and the end game. This hopefully eliminates any crashes to the ready prompt.
* The avoid combat trainer now works in dungeons too.


## v2.1 (2015-04-09)

* Saving to EasyFlash cartridges that don't use Am29F040 chips (including Chameleon) now works.
* Exit solo mode if character leaves combat.
* [Don't spawn a ship](/2015/04/02/u4-dungeon-frigate.html) when pressing B while standing on a ladder in a dungeon.
* Joshua courage -> canrage typo.
* Fixed east/west start positions in Destard rooms on level 8.
* Removed glitch when attacking sick people in townes.
* Play correct music after dying and being resurrected by Lord British.
* Fixed wind status after loading a game in a dungeon and then exiting.
* Clarified the recommendation for emulator users when flipping d64s.
* Proper error message in editor if no savegame is found in cartridge.


## v2.0 (2015-03-29)

### Graphics Overhaul

* Character creation intro has new graphics drawn by Vanja 'Mermaid' Utne.
* The tileset has been updated with new and more colorful tiles, mainly backported from Ultima V.
* When peering at a gem or casting view the map is now in color.
* The dungeon renderer has been sped up, and items and monsters are drawn in color.

### Text Overhaul

* NPC dialogue and game text now uses mixed case.
* Many dialogue bugs have been fixed:
    * Fixed broken question trigger and added missing dialogue for Thevel in Britain.
    * Added missing dialogue for Serpent's Hold gate guards.
    * Changed keyword for Michelle to avoid conflict with health in Serpent's Hold.
    * Fixed broken message trigger for Water in Castle British.
    * Fixed misspelled keyword for Estro in the Lycaeum.
    * Fixed misspelled keyword for a poor beggar in Yew.
    * Fixed broken message trigger for Alkerion in Minoc.
    * Fixed broken message trigger for Shamino in Skara Brae.

src/talk/talk.s
	.include "uscii.i"


current_location	= $1a
party_size		= $1f
last_humility_check	= $27
move_counter		= $2c
console_xpos		= $4e
currplayer		= $54
bcdnum			= $57
towne_virtue		= $55
zptmp1			= $5a
lt_x			= $61
monster_type		= $66
ptr2			= $7c
ptr1			= $7e


j_waitkey		= $0800
j_primm 		= $0821
j_console_out		= $0824
j_get_stats_ptr 	= $082d
j_update_status 	= $0845
j_update_view		= $084b
j_rand			= $084e
j_getnumber		= $085a
j_clearstatwindow	= $0869

party_stats		= $aa00

game_stats		= $ab00
gold			= $ab13

map_status		= $ac00
object_tile		= $ac60

inbuffer		= $af00


	.segment "DIALOG"

dialog_strings:			.res $f5
dialog_keyword_1:		.res 4
dialog_keyword_2:		.res 4
dialog_question_trigger:	.res 1
dialog_humility:		.res 1
dialog_turns_away_prob:		.res 1


	.segment "TALK"

talk:
	ldx	#$03
@copy_keywords:
	lda	dialog_keyword_1,x
	ora	#$80
	sta	keyword_1,x
	lda	dialog_keyword_2,x
	ora	#$80
	sta	keyword_2,x
	dex
	bpl	@copy_keywords
	ldx	$6A
	lda	object_tile,x
	sta	monster_type
	jsr	j_primm
	.byte $8d
	.byte "You meet", $8d
	.byte 0

	lda	#$03
	jsr	print_string
	lda	#$8D
	jsr	j_console_out
	jsr	j_rand
	cmp	dialog_turns_away_prob
	bcc	maybe_fight
	jsr	j_rand
	bmi	talk_prompt
	lda	#$8D
	jsr	j_console_out
print_name:
	jsr	pronoun_says
	jsr	j_primm
	.byte " I am", $8d
	.byte 0

	lda	#$01
	jsr	print_string
	lda #$ae
	jsr j_console_out
	lda	#$8D
	jsr	j_console_out
	jmp	talk_prompt

maybe_fight:
	sta	$58
	lda	dialog_turns_away_prob
	sec
	sbc	$58
	cmp	#$40
	bcs	@fight
	jmp	pronoun_turns_away

@fight:
	jmp	start_combat

; dead code, unused in original game
;	jsr	pronoun_says_newline
;	jsr	j_primm
;	.byte "Go away!", $8d
;	.byte 0

talk_prompt:
	jsr	j_primm
	.byte $8d
	.byte "Your interest:", $8d
	.byte 0

	jsr	get_input
	lda	#$8D
	jsr	j_console_out
	jsr	j_rand
	cmp	dialog_turns_away_prob
	bcs	@check_input
	sta	$58
	sec
	lda	dialog_turns_away_prob
	sbc	$58
	cmp	#$80
	bcs	@fight
	jmp	pronoun_turns_away

@fight:
	jmp	start_combat

@check_input:
	jsr	compare_keywords
	bpl	@found_keyword
	jsr	pronoun_says
	jsr	j_primm
	.byte " That,", $8d
	.byte "I cannot help", $8d
	.byte "thee with.", $8d
	.byte 0

	jmp	check_question_trigger

@found_keyword:
	cmp	#$00
	beq	@bye
	cmp	#$09
	bne	@name
@bye:
	jsr	pronoun_says
	jsr	j_primm
	.byte " Bye.", $8d
	.byte 0

	jmp	talk_done

@name:
	cmp	#$01
	bne	@look
	jmp	print_name

@look:
	cmp	#$02
	bne	@join
	jsr	j_primm
	.byte "You see a", $8d
	.byte 0

	lda	#$03
	jmp	print_response_newline

@join:
	cmp	#$07
	bne	@give
	jmp	check_join

@give:
	cmp	#$08
	bne	@keyword
	jmp	check_give

@keyword:
	jsr	pronoun_says_newline
	lda	$6A
	clc
	adc	#$01
print_response_newline:
	jsr	print_string
	lda	#$8D
	jsr	j_console_out
	jmp	check_question_trigger

check_question_trigger:
	lda	$6A
	cmp	dialog_question_trigger
	beq	ask_question
	jmp	talk_prompt

ask_question:
	jsr	j_waitkey
	lda	#$8D
	jsr	j_console_out
	lda	#$02
	jsr	print_string
	jsr	j_primm
	.byte " asks:", $8d
	.byte 0

	lda	#$08
	jsr	print_string
@get_response:
	jsr	j_primm
	.byte $8d
	.byte $8d
	.byte "You respond:", $8d
	.byte 0

	jsr	get_input
	lda	#$8D
	jsr	j_console_out
	lda	inbuffer
	cmp	#$CE
	beq	@no
	cmp	#$D9
	bne	@yes_or_no
	jsr	subtract_humility
	jsr	pronoun_says_newline
	lda	#$09
	sta	$6A
	jmp	print_response_newline

@no:
	jsr	add_humility
	jsr	pronoun_says_newline
	lda	#$0A
	sta	$6A
	jmp	print_response_newline

@yes_or_no:
	jsr	pronoun_says_newline
	jsr	j_primm
	.byte "Yes, or no:", 0

	jmp	@get_response

pronoun_turns_away:
	lda	#$02
	jsr	print_string
	jsr	j_primm
	.byte " turns away!", $8d
	.byte $8d
	.byte 0

	jmp	talk_done

start_combat:
	lda	#$02
	jsr	print_string
	jsr	j_primm
	.byte " says:", $8d
	.byte "En garde! Fool!", $8d
	.byte 0

	ldx	$6A
	lda	#$FF
	sta	$ACC0,x
	jmp	talk_done

pronoun_says:
	lda	#$02
	jsr	print_string
	jsr	j_primm
	.byte " says:", 0

	rts

pronoun_says_newline:
	jsr	pronoun_says
	lda	#$8D
	jsr	j_console_out
	rts

get_input:
	lda	#$BF				; ?
	jsr	j_console_out
	lda	#$00
	sta	input_ctr
@get_char:
	jsr	j_waitkey
	cmp	#$8D
	beq	@got_input
	cmp	#$94
	beq	@backspace
	cmp	#$A0
	bcc	@get_char
	ldx	input_ctr
	sta	inbuffer,x
	jsr	j_console_out
	inc	input_ctr
	lda	input_ctr
	cmp	#$0F
	bcc	@get_char
	bcs	@got_input
@backspace:
	lda	input_ctr
	beq	@get_char
	dec	input_ctr
	dec	console_xpos
	lda	#$A0
	jsr	j_console_out
	dec	console_xpos
	jmp	@get_char

@got_input:
	ldx	input_ctr
	lda	#$A0
@pad_spaces:
	sta	inbuffer,x
	inx
	cpx	#$06
	bcc	@pad_spaces
	lda	#$8D
	jsr	j_console_out
	rts

input_ctr:
	.byte 0

compare_keywords:
	lda	#$09
	sta	$6A
@next:
	lda	$6A
	asl	a
	asl	a
	tay
	ldx	#$00
@compare:
	lda	keywords,y
	cmp	inbuffer,x
	bne	@nomatch
	iny
	inx
	cpx	#$04
	bcc	@compare
	lda	$6A
	rts

@nomatch:
	dec	$6A
	bpl	@next
	lda	$6A
	rts

keywords:
	.byte "BYE "
@name:
	.byte "NAME"
@look:
	.byte "LOOK"
@job:
	.byte "JOB "
@heal:
	.byte "HEAL"
keyword_1:
	.byte "    "
keyword_2:
	.byte "    "
@join:
	.byte "JOIN"
@give:
	.byte "GIVE"

	.byte "    "

print_string:
	tax
	lda #<dialog_strings
	sta ptr1
	lda #>dialog_strings
	sta ptr1 + 1
@next:
	dex
	beq @print
	ldy #0
@check:
	lda (ptr1),y
	bpl @eos
	iny
	bne @check
@eos:
	tya
	sec
	adc ptr1
	sta ptr1
	bne @next
@print:
	ldy #0
	lda (ptr1),y
	bpl @last
	jsr j_console_out
	inc ptr1
	bne @print
@last:
	ora #$80
	jmp j_console_out


check_join:
	lda	lt_x
	bne	@cant_join
	lda	current_location
	sec
	sbc	#$05
	cmp	#$08
	bcs	@cant_join
	sta	towne_virtue
	lda	#$01
	sta	currplayer
	jsr	j_get_stats_ptr
	ldy	#$11
	lda	(ptr1),y
	cmp	towne_virtue
	beq	@cant_join
	ldx	towne_virtue
	lda	game_stats,x
	beq	@part_avatar
	cmp	#$40
	bcc	@decline
@part_avatar:
	lda	party_size
	cmp	party_stats + 26
	bcs	@not_enough_xp
	jmp	accept_join

@cant_join:
	jsr	pronoun_says_newline
	jsr	j_primm
	.byte "I cannot", $8d
	.byte "join thee.", $8d
	.byte 0

	jmp	talk_prompt

@decline:
	stx	$58
@print_decline:
	jsr	pronoun_says_newline
	jsr	j_primm
	.byte "Thou art not", $8d
	.byte 0

	jsr	decline_reason
	jsr	j_primm
	.byte $8d
	.byte "enough for me", $8d
	.byte "to join thee.", $8d
	.byte 0

	jmp	talk_prompt

@not_enough_xp:
	lda	#$08
	sta	$58
	jmp	@print_decline

accept_join:
	jsr	pronoun_says_newline
	jsr	j_primm
	.byte "I am honored", $8d
	.byte "to join thee!", $8d
	.byte 0

	ldx	#$1F
	lda	#$00
	sta	object_tile,x
	sta	map_status,x
	sta	$ACC0,x
	sta	$ACE0,x
	jsr	j_update_view
	lda	#$08
	sta	currplayer
@next:
	jsr	j_get_stats_ptr
	ldy	#$11
	lda	(ptr1),y
	cmp	towne_virtue
	beq	@found_slot
	dec	currplayer
	bpl	@next
	jmp	@done

@found_slot:
	jsr	j_get_stats_ptr
	lda	ptr1
	sta	ptr2
	lda	ptr1 + 1
	sta	ptr2 + 1
	inc	party_size
	lda	party_size
	sta	currplayer
	jsr	j_get_stats_ptr
	ldy	#$1F
@move:
	lda	(ptr1),y
	pha
	lda	(ptr2),y
	sta	(ptr1),y
	pla
	sta	(ptr2),y
	dey
	bpl	@move
	jsr	j_clearstatwindow
	jsr	j_update_status
@done:
	jmp	talk_done

check_give:
	lda	monster_type
	cmp	#$58
	beq	@how_much
	jsr	pronoun_says
	jsr	j_primm
	.byte " I do", $8d
	.byte "not need thy", $8d
	.byte "gold, keep it!", $8d
	.byte 0

	jmp	check_question_trigger

@how_much:
	jsr	j_primm
	.byte "How much-", 0

	jsr	j_getnumber
	lda	#$8D
	jsr	j_console_out
	lda	bcdnum
	bne	@give_amount
	jmp	check_question_trigger

@give_amount:
	jsr	subtract_gold
	bpl	@has_gold
	jsr	j_primm
	.byte $8d
	.byte "Thou hast not", $8d
	.byte "that much gold!", $8d
	.byte 0

	jmp	check_question_trigger

@has_gold:
	jsr	give_gold_add_virtue
	lda	#$8D
	jsr	j_console_out
	jsr	pronoun_says
	jsr	j_primm
	.byte " Oh,", $8d
	.byte "thank thee!", $8d
	.byte "I shall never", $8d
	.byte "forget thy", $8d
	.byte "kindness!", $8d
	.byte 0

	jmp	check_question_trigger

give_gold_add_virtue:
	lda	gold + 1
	ora	gold
	bne	@has_gold_left
	lda	#$03
	ldy	#$04  ; cycle 1 at shrine of SACRIFICE recommends this action. Original code awarded HONOR (#$05).
	jsr	add_virtue
@has_gold_left:
	lda	move_counter + 3
	and	#$F0
	cmp	last_humility_check
	beq	@too_soon
	sta	last_humility_check
	lda	#$02
	ldy	#$01
	jsr	add_virtue
@too_soon:
	rts

decline_reason:
	ldx	$58
	bne	@compassionate
	jsr	j_primm
	.byte "honest", 0

	rts

@compassionate:
	dex
	bne	@valiant
	jsr	j_primm
	.byte "compassionate", 0

	rts

@valiant:
	dex
	bne	@just
	jsr	j_primm
	.byte "valiant", 0

	rts

@just:
	dex
	bne	@sacrificial
	jsr	j_primm
	.byte "just", 0

	rts

@sacrificial:
	dex
	bne	@honorable
	jsr	j_primm
	.byte "sacrificial", 0

	rts

@honorable:
	dex
	bne	@spiritual
	jsr	j_primm
	.byte "honorable", 0

	rts

@spiritual:
	dex
	bne	@humble
	jsr	j_primm
	.byte "spiritual", 0

	rts

@humble:
	dex
	bne	@experienced
	jsr	j_primm
	.byte "humble", 0

	rts  ; BUG FIX, this was missing in original code

@experienced:
	jsr	j_primm
	.byte "experienced", 0

	rts

subtract_humility:
	lda	dialog_humility
	cmp	#$01
	bne	@not_humility
	lda	#$05
	ldy	#$07
	jsr	subtract_virtue
	lda	move_counter + 3
	and	#$F0
	sta	last_humility_check
@not_humility:
	rts

add_humility:
	lda	dialog_humility
	cmp	#$01
	bne	@not_humility
	lda	move_counter + 3
	and	#$F0
	cmp	last_humility_check
	beq	@not_humility
	sta	last_humility_check
	lda	#$0A
	ldy	#$07
	jsr	add_virtue
@not_humility:
	rts

subtract_gold:
	sed
	sec
	lda	gold + 1
	sbc	bcdnum
	sta	gold + 1
	lda	gold
	sbc	#$00
	bcc	@underflow
	sta	gold
	cld
	jsr	j_update_status
	lda	#$00
	rts

@underflow:
	clc
	lda	gold + 1
	adc	bcdnum
	sta	gold + 1
	cld
	lda	#$FF
	rts

add_virtue:
	sta	$59
	sed
	clc
	lda	game_stats,y
	beq	@overflow
	adc	$59
	bcc	@overflow
	lda	#$99
@overflow:
	sta	game_stats,y
	cld
	rts

subtract_virtue:
	sta	zptmp1
	sty	$59
	lda	game_stats,y
	beq	@lost_an_eigth
@set_virtue:
	sed
	sec
	sbc	zptmp1
	bcs	@underflow
	bne	@underflow
	lda	#$01
@underflow:
	sta	game_stats,y
	cld
	rts

@lost_an_eigth:
	jsr	j_primm
	.byte $8d
	.byte "Thou hast lost", $8d
	.byte "an eighth!", $8d
	.byte 0

	ldy	$59
	lda	#$99
	jmp	@set_virtue

talk_done:
	rts
